File: 110-balanced-binary-tree/110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root: return 0
        return 1 + max(self.maxDepth(root.left), self.maxDepth(root.right))
    
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root: return True
        if abs(self.maxDepth(root.left) - self.maxDepth(root.right)) > 1 or not self.isBalanced(root.left) or not self.isBalanced(root.right):
                return False
        return True
        

File: 110-balanced-binary-tree/110-balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private static int maxDepth(TreeNode root) {
        return root == null ? 0 : 1 + Math.max(maxDepth(root.left), maxDepth(root.right));
    }
    public boolean isBalanced(TreeNode root) {
        if(root == null) return true;
        if((Math.abs(maxDepth(root.left) - maxDepth(root.right)) > 1)
          || (!isBalanced(root.left) || !isBalanced(root.right)))
            return false;
        return true;
    }
}